File: src/domain/story.ts
import { createToyBaseObject, type BaseObject, type IsoDateTime, type StableId } from "./common";
import type { ContextTaskType } from "./context-pack";

export interface Work extends BaseObject {
  type: "work";
  workKind: "novel" | "series" | "short" | "other";
  premise: string;
  readerContractRefs: StableId[];
  outline: WorkOutline;
}

export interface WorkOutline {
  id: StableId;
  workId: StableId;
  volumes: VolumeOutline[];
}

export interface VolumeOutline {
  id: StableId;
  workId: StableId;
  title: string;
  sequence: number;
  purpose: string;
  segments: StorySegment[];
}

export interface StorySegment {
  id: StableId;
  workId: StableId;
  parentId?: StableId;
  label: string;
  sequence: number;
  goal: string;
  pressure: string;
  chapters: ChapterPlan[];
}

export interface ChapterPlan {
  id: StableId;
  workId: StableId;
  sequence: number;
  title: string;
  chapterGoal: string;
  sceneRefs: StableId[];
}

export interface Beat {
  id: StableId;
  sceneId: StableId;
  sequence: number;
  beatKind: "action" | "dialogue" | "reveal" | "decision" | "reversal" | "transition" | "other";
  intent: string;
  expectedChange: string;
  characterRefs: StableId[];
  evidenceRefIds: StableId[];
}

export interface SceneCard extends BaseObject {
  type: "scene-card";
  chapterId: StableId;
  sceneGoal: string;
  povCharacterId?: StableId;
  povMode: string;
  knowledgeBoundaryRef: StableId;
  settingPressureRef?: StableId;
  settingPressureText?: string;
  entryState: string;
  exitState: string;
  characterStateRefs: StableId[];
  relationshipRefs: StableId[];
  foreshadowingRefs: StableId[];
  styleTargetRefs: StableId[];
  beats: Beat[];
}

export type StorySceneCard = SceneCard;

export interface ForeshadowingEntry {
  id: StableId;
  ledgerId: StableId;
  signal: string;
  plantedAtRef: StableId;
  intendedPayoff: string;
  currentState: "planted" | "partially-revealed" | "resolved" | "abandoned";
  visibility: "hidden" | "subtle" | "clear";
  riskNotes: string[];
}

export interface ForeshadowingLedger extends BaseObject {
  type: "foreshadowing-ledger";
  workId: StableId;
  entries: ForeshadowingEntry[];
  lastReviewedAt: IsoDateTime;
}

export interface RevisionLayer extends BaseObject {
  type: "revision-layer";
  targetRef: StableId;
  layerKind: "structure" | "character" | "scene" | "prose-style" | "continuity";
  objective: string;
  reviewItemRefs: StableId[];
}

export interface ContinuityIssue extends BaseObject {
  type: "continuity-issue";
  targetRef: StableId;
  issueKind: string;
  description: string;
  relatedRefs: StableId[];
  severity: "low" | "medium" | "high";
  proposedFix: string;
  confirmationState: "draft" | "accepted" | "rejected";
}

export interface DraftReviewItem extends BaseObject {
  type: "draft-review-item";
  targetRef: StableId;
  sourceDraftRef: StableId;
  severity: "low" | "medium" | "high";
  category: string;
  finding: string;
  recommendation: string;
  evidenceRefIds: StableId[];
  blockedReason?: string;
  conflictRefs?: StableId[];
}

export interface RevisionRequest extends BaseObject {
  type: "revision-request";
  targetRef: StableId;
  originalChunkRef: StableId;
  originalDraftRef: StableId;
  userFeedback: string;
  outputSchemaRef: string;
  outputSchemaVersion: number;
  contextHash: string;
  requestedRerunTaskType: ContextTaskType;
  requestStatus: "draft" | "ready-for-rerun" | "rerun-created" | "blocked" | "closed";
}

export interface ChangeSetEntry {
  targetRef: StableId;
  changeType: "create" | "update" | "remove";
  description: string;
}

export interface ChangeSet extends BaseObject {
  type: "change-set";
  summary: string;
  reason: string;
  sourceDraftRef: StableId;
  revisionRequestRefs: StableId[];
  targetRefs: StableId[];
  changes: ChangeSetEntry[];
  impactScope: string[];
  confirmationState: "pending-user-confirmation" | "confirmed" | "rejected";
}

export interface RevisionSuggestion {
  targetRef: StableId;
  suggestedTextRef: StableId;
  reason: string;
  evidenceRefIds: StableId[];
  riskNotes: string[];
}

export interface ProseQualityPass extends BaseObject {
  type: "prose-quality-pass";
  targetRef: StableId;
  qualityGoals: string[];
  beforeRef: StableId;
  suggestions: RevisionSuggestion[];
  decision: "pending-user-review" | "accepted" | "rejected";
}

export interface StyleRevisionPass extends BaseObject {
  type: "style-revision-pass";
  targetRef: StableId;
  styleProfileRefs: StableId[];
  voiceConstraints: string[];
  suggestions: RevisionSuggestion[];
  decision: "pending-user-review" | "accepted" | "rejected";
}

export function createToyWork(timestamp: IsoDateTime): Work {
  return {
    ...createToyBaseObject("work", "work-public-toy", "Public Toy Work", timestamp),
    workKind: "novel",
    premise: "A learner must choose which clue to preserve under a visible cost.",
    readerContractRefs: ["reader-contract-toy"],
    outline: {
      id: "outline-public-toy",
      workId: "work-public-toy",
      volumes: [],
    },
  };
}

export function createToySceneCard(timestamp: IsoDateTime): SceneCard {
  return {
    ...createToyBaseObject("scene-card", "scene-toy-gate", "Toy Gate Choice", timestamp),
    chapterId: "chapter-toy-one",
    sceneGoal: "Force a costed choice.",
    povMode: "third-person-limited",
    knowledgeBoundaryRef: "knowledge-boundary-toy",
    settingPressureText: "A token opens only one route.",
    entryState: "The clue path is unknown.",
    exitState: "One route is chosen and one is closed.",
    characterStateRefs: ["dynamic-state-gate-choice"],
    relationshipRefs: ["relationship-mentor-learner"],
    foreshadowingRefs: ["foreshadowing-token-cost"],
    styleTargetRefs: ["style-profile-toy"],
    beats: [
      {
        id: "beat-token-decision",
        sceneId: "scene-toy-gate",
        sequence: 1,
        beatKind: "decision",
        intent: "Make the resource cost explicit.",
        expectedChange: "The protagonist commits to one route.",
        characterRefs: ["character-practical-learner"],
        evidenceRefIds: ["evidence-toy-pressure"],
      },
    ],
  };
}

export function createToyRevisionRequest(timestamp: IsoDateTime): RevisionRequest {
  return {
    ...createToyBaseObject(
      "revision-request",
      "revision-request-toy",
      "Toy revision request",
      timestamp,
    ),
    targetRef: "evidence-toy-pressure",
    originalChunkRef: "source-chunk-toy",
    originalDraftRef: "draft-output-fake-toy",
    userFeedback: "Clarify the cost before the choice.",
    outputSchemaRef: "toy-draft-output",
    outputSchemaVersion: 1,
    contextHash: "0000000000000000000000000000000000000000000000000000000000000000",
    requestedRerunTaskType: "source-breakdown",
    requestStatus: "ready-for-rerun",
  };
}

export function createToyProseQualityPass(timestamp: IsoDateTime): ProseQualityPass {
  return {
    ...createToyBaseObject(
      "prose-quality-pass",
      "prose-quality-pass-toy",
      "Toy prose quality pass",
      timestamp,
    ),
    targetRef: "scene-toy-gate",
    qualityGoals: ["clarity", "cause-before-effect"],
    beforeRef: "draft-before-toy",
    suggestions: [
      {
        targetRef: "draft-before-toy",
        suggestedTextRef: "draft-after-toy",
        reason: "Improve clarity before the costed choice.",
        evidenceRefIds: ["evidence-toy-pressure"],
        riskNotes: ["Do not imitate source phrasing."],
      },
    ],
    decision: "pending-user-review",
  };
}

export function createToyChangeSet(timestamp: IsoDateTime): ChangeSet {
  return {
    ...createToyBaseObject("change-set", "change-set-toy", "Toy change set", timestamp),
    summary: "Clarify the toy gate cost.",
    reason: "User accepted a revision suggestion.",
    sourceDraftRef: "draft-output-fake-toy",
    revisionRequestRefs: ["revision-request-toy"],
    targetRefs: ["scene-toy-gate"],
    changes: [
      {
        targetRef: "scene-toy-gate",
        changeType: "update",
        description: "Make the cost explicit before the choice.",
      },
    ],
    impactScope: ["scene"],
    confirmationState: "pending-user-confirmation",
  };
}

File: crates/core/src/model/story.rs
use serde::{Deserialize, Serialize};

use super::common::{BaseObject, IsoDateTime, StableId};
use super::context_pack::ContextTaskType;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Work {
    #[serde(flatten)]
    pub base: BaseObject,
    pub work_kind: String,
    pub premise: String,
    pub reader_contract_refs: Vec<StableId>,
    pub outline: WorkOutline,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkOutline {
    pub id: StableId,
    pub work_id: StableId,
    pub volumes: Vec<VolumeOutline>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VolumeOutline {
    pub id: StableId,
    pub work_id: StableId,
    pub title: String,
    pub sequence: u32,
    pub purpose: String,
    pub segments: Vec<StorySegment>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StorySegment {
    pub id: StableId,
    pub work_id: StableId,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parent_id: Option<StableId>,
    pub label: String,
    pub sequence: u32,
    pub goal: String,
    pub pressure: String,
    pub chapters: Vec<ChapterPlan>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChapterPlan {
    pub id: StableId,
    pub work_id: StableId,
    pub sequence: u32,
    pub title: String,
    pub chapter_goal: String,
    pub scene_refs: Vec<StableId>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Beat {
    pub id: StableId,
    pub scene_id: StableId,
    pub sequence: u32,
    pub beat_kind: String,
    pub intent: String,
    pub expected_change: String,
    pub character_refs: Vec<StableId>,
    pub evidence_ref_ids: Vec<StableId>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SceneCard {
    #[serde(flatten)]
    pub base: BaseObject,
    pub chapter_id: StableId,
    pub scene_goal: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pov_character_id: Option<StableId>,
    pub pov_mode: String,
    pub knowledge_boundary_ref: StableId,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub setting_pressure_ref: Option<StableId>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub setting_pressure_text: Option<String>,
    pub entry_state: String,
    pub exit_state: String,
    pub character_state_refs: Vec<StableId>,
    pub relationship_refs: Vec<StableId>,
    pub foreshadowing_refs: Vec<StableId>,
    pub style_target_refs: Vec<StableId>,
    pub beats: Vec<Beat>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ForeshadowingEntry {
    pub id: StableId,
    pub ledger_id: StableId,
    pub signal: String,
    pub planted_at_ref: StableId,
    pub intended_payoff: String,
    pub current_state: String,
    pub visibility: String,
    pub risk_notes: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ForeshadowingLedger {
    #[serde(flatten)]
    pub base: BaseObject,
    pub work_id: StableId,
    pub entries: Vec<ForeshadowingEntry>,
    pub last_reviewed_at: IsoDateTime,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RevisionLayer {
    #[serde(flatten)]
    pub base: BaseObject,
    pub target_ref: StableId,
    pub layer_kind: String,
    pub objective: String,
    pub review_item_refs: Vec<StableId>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ContinuityIssue {
    #[serde(flatten)]
    pub base: BaseObject,
    pub target_ref: StableId,
    pub issue_kind: String,
    pub description: String,
    pub related_refs: Vec<StableId>,
    pub severity: String,
    pub proposed_fix: String,
    pub confirmation_state: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DraftReviewItem {
    #[serde(flatten)]
    pub base: BaseObject,
    pub target_ref: StableId,
    pub source_draft_ref: StableId,
    pub severity: String,
    pub category: String,
    pub finding: String,
    pub recommendation: String,
    pub evidence_ref_ids: Vec<StableId>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub blocked_reason: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub conflict_refs: Option<Vec<StableId>>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RevisionRequest {
    #[serde(flatten)]
    pub base: BaseObject,
    pub target_ref: StableId,
    pub original_chunk_ref: StableId,
    pub original_draft_ref: StableId,
    pub user_feedback: String,
    pub output_schema_ref: String,
    pub output_schema_version: u16,
    pub context_hash: String,
    pub requested_rerun_task_type: ContextTaskType,
    pub request_status: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChangeSetEntry {
    pub target_ref: StableId,
    pub change_type: String,
    pub description: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChangeSet {
    #[serde(flatten)]
    pub base: BaseObject,
    pub summary: String,
    pub reason: String,
    pub source_draft_ref: StableId,
    pub revision_request_refs: Vec<StableId>,
    pub target_refs: Vec<StableId>,
    pub changes: Vec<ChangeSetEntry>,
    pub impact_scope: Vec<String>,
    pub confirmation_state: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RevisionSuggestion {
    pub target_ref: StableId,
    pub suggested_text_ref: StableId,
    pub reason: String,
    pub evidence_ref_ids: Vec<StableId>,
    pub risk_notes: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProseQualityPass {
    #[serde(flatten)]
    pub base: BaseObject,
    pub target_ref: StableId,
    pub quality_goals: Vec<String>,
    pub before_ref: StableId,
    pub suggestions: Vec<RevisionSuggestion>,
    pub decision: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StyleRevisionPass {
    #[serde(flatten)]
    pub base: BaseObject,
    pub target_ref: StableId,
    pub style_profile_refs: Vec<StableId>,
    pub voice_constraints: Vec<String>,
    pub suggestions: Vec<RevisionSuggestion>,
    pub decision: String,
}
